pkg/controller/clusterpool/clusterpool_controller.go

```diff
@@ -0,0 +1,389 @@
+	var credsSecretName string
+	if clp.Spec.Platform.AWS != nil {
+		credsSecretName = clp.Spec.Platform.AWS.CredentialsSecretRef.Name
+		// Lookup the platform creds secret for this pool which we are currently assuming
+		// is in the hive namespace:
+		credsSecret := &corev1.Secret{}
+		if err := r.Client.Get(context.Background(), types.NamespacedName{Namespace: controllerutils.GetHiveNamespace(), Name: credsSecretName}, credsSecret); err != nil {
+			logger.WithError(err).Error("error looking up credentials secret for pool in hive namespace")
+			return err
+		}
+
+		accessKeyID := credsSecret.Data[constants.AWSAccessKeyIDSecretKey]
+		secretAccessKey := credsSecret.Data[constants.AWSSecretAccessKeySecretKey]
+		awsProvider := &clusterresource.AWSCloudBuilder{
+			AccessKeyID:     string(accessKeyID),
+			SecretAccessKey: string(secretAccessKey),
+		}
+		builder.CloudBuilder = awsProvider
+
+	} else if clp.Spec.Platform.GCP != nil {
+		credsSecretName = clp.Spec.Platform.GCP.CredentialsSecretRef.Name
+		// Lookup the platform creds secret for this pool which we are currently assuming
+		// is in the hive namespace:
+		credsSecret := &corev1.Secret{}
+		if err := r.Client.Get(context.Background(), types.NamespacedName{Namespace: controllerutils.GetHiveNamespace(), Name: credsSecretName}, credsSecret); err != nil {
+			logger.WithError(err).Error("error looking up credentials secret for pool in hive namespace")
+			return err
+		}
+
+		gcpSA := credsSecret.Data[constants.GCPCredentialsName]
+
+		projectID, err := gcpclient.ProjectID(gcpSA)
+		if err != nil {
+			return errors.Wrap(err, "error loading GCP project ID from service account json")
+		}
+
+		gcpProvider := &clusterresource.GCPCloudBuilder{
+			ServiceAccount: gcpSA,
+			ProjectID:      projectID,
+		}
+		builder.CloudBuilder = gcpProvider
+
+	} else if clp.Spec.Platform.Azure != nil {
+		credsSecretName = clp.Spec.Platform.Azure.CredentialsSecretRef.Name
+		// Lookup the platform creds secret for this pool which we are currently assuming
+		// is in the hive namespace:
+		credsSecret := &corev1.Secret{}
+		if err := r.Client.Get(context.Background(), types.NamespacedName{Namespace: controllerutils.GetHiveNamespace(), Name: credsSecretName}, credsSecret); err != nil {
+			logger.WithError(err).Error("error looking up credentials secret for pool in hive namespace")
+			return err
+		}
+		azureSP := credsSecret.Data[constants.AzureCredentialsName]
+
+		azureProvider := &clusterresource.AzureCloudBuilder{
+			ServicePrincipal:            azureSP,
+			BaseDomainResourceGroupName: clp.Spec.Platform.Azure.BaseDomainResourceGroupName,
+		}
+		builder.CloudBuilder = azureProvider
+	}
+	// TODO: OpenStack, VMware, and Ovirt.
+
+	objs, err := builder.Build()
+	if err != nil {
+		return errors.Wrap(err, "error building resources")
+	}
+	for _, obj := range objs {
+		if err := r.Client.Create(context.Background(), obj); err != nil {
```

Comment:
I remember from our discussion that we wanted to have the ClusterDeployment be the last thing created to ensure the ClusterDeployment does not attempt to install when there were errors laying down other dependent resources.

Reply:
Can you refresh my memory on this, I can't recall the details of that convo.

Should we just have the builder return that resource last?

Reply:
Let's say that the creds secret is created after the ClusterDeployment is created. Now, let's say that there is an error creating the creds secret. The ClusterDeployment will be stuck in the installing state, counting against the pool's count but never available to be used. If the ClusterDeployment is created last, then it will only be created if all its dependent resources were also created. If there is an error along the way, everything will be cleaned up by the namespace reaper since the namespace will not contain a ClusterDeployment.

Reply:
Ok making it created last sounds cleaner than refactoring all this to return the CD separately, I'd hoped to keep that builder object just returning an arbitrary list of resources to create. I will move the CD return to last.

If anyone is still using int we should probably address that underlying rbac problem.

Reply:
You do not need to change the builder at all. Just have this controller move the CD to the end of the slice.